### PR TITLE
Integrate browserstack (do not merge without discussing with me)

### PR DESCRIPTION
**Pull request by oskarpearson**
Comments please!

(Note that this if you run this, you will get CSRF errors on logout. I'm not sure if that's a real problem, or if it's caused by running tests in parallel. Please don't merge this until we're sure.)

Added a small selection of browsers to the browser.json list. For a full list, see
  `curl -u "user:password" https://www.browserstack.com/automate/browsers.json`

We need to decide which browsers to use from the above json file, since if we
match the full set defined in the requirements things will take a very very long
time to run.
- Integrated Rake task to run browserstack tests in parallel
- Made the docker-compose container names depend on the BUILD_NUMBER, for running under CI concurrently
- Integrated the BrowserStackLocal binary into the repo, downloaded from
  https://www.browserstack.com/automate/ruby (Log in -> Docs -> Ruby -> Setting up Local Testing -> Linux 64-bit)

Others:
- Corrected the name of our app, since it's changed in one of the other repos
- Changed the order of our sleeps and prints, so that in the best case no extraneous '.' chars are printed
- Corrected some typos in the README
- Changed some of the ' characters in the code to " characters, to match our project coding standards

### CODE REVIEW

**Comment by oskarpearson** [comment]
@ansonK This should be ready for your comments.

File: spec/support/failed_session.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+class FailedSession
+   include Singleton
```
Comment (FriedSock) [inline spec/support/failed_session.rb:2]: Didn't know about `Singleton`, pretty cool

**Comment by oskarpearson** [comment]
@ansonK Could you review this when you have a chance? I'd like to close the associated Pivotal task.

Thanks!

Oskar

**Comment by ansonK** [comment]
Looks good. I did get a test failure for IE7 tho. Was this expected?

```
Failures:

  1) Authentication Can log in to the service app, using the auth app
     Failure/Error: we_are_back_on_the_service_app
       expected no Exception, got #<Selenium::WebDriver::Error::StaleElementReferenceError: Element is no longer valid (WARNING: The server did not provide any stacktrace information)
       Command duration or timeout: 16 milliseconds
       For documentation on this error, please visit: http://seleniumhq.org/exceptions/stale_element_reference.html
       Build info: version: '2.43.1', revision: '5163bce', time: '2014-09-10 16:27:33'
       System info: host: '185-44-128-200', ip: '185.44.128.200', os.name: 'windows', os.arch: 'x86', os.version: '5.2', java.version: '1.8.0_40'
       Driver info: org.openqa.selenium.ie.InternetExplorerDriver
       Capabilities [{browserAttachTimeout=0, enablePersistentHover=true, ie.forceCreateProcessApi=false, ie.usePerProcessProxy=false, ignoreZoomSetting=false, handlesAlerts=true, version=7, platform=WINDOWS, nativeEvents=true, ie.ensureCleanSession=true, elementScrollBehavior=0, ie.browserCommandLineSwitches=, requireWindowFocus=false, browserName=internet explorer, initialBrowserUrl=http://localhost:24958/, takesScreenshot=true, javascriptEnabled=true, ignoreProtectedModeSettings=false, enableElementCacheCleanup=true, cssSelectorsEnabled=true, unexpectedAlertBehaviour=dismiss}]
       Session ID: af6801e4-63da-4729-9234-0bd63be1ed60 (org.openqa.selenium.StaleElementReferenceException)> with backtrace:
         # [remote server] sun.reflect.NativeConstructorAccessorImpl():-2:in `newInstance0'
         # [remote server] sun.reflect.NativeConstructorAccessorImpl():-1:in `newInstance'
         # [remote server] sun.reflect.DelegatingConstructorAccessorImpl():-1:in `newInstance'
         # [remote server] java.lang.reflect.Constructor():-1:in `newInstance'
         # [remote server] org.openqa.selenium.remote.ErrorHandler(ErrorHandler.java):204:in `createThrowable'
         # [remote server] org.openqa.selenium.remote.ErrorHandler(ErrorHandler.java):156:in `throwIfResponseFailed'
         # [remote server] org.openqa.selenium.remote.RemoteWebDriver(RemoteWebDriver.java):599:in `execute'
         # [remote server] org.openqa.selenium.remote.RemoteWebElement(RemoteWebElement.java):268:in `execute'
         # [remote server] org.openqa.selenium.remote.RemoteWebElement(RemoteWebElement.java):152:in `getText'
         # [remote server] sun.reflect.GeneratedMethodAccessor4():-1:in `invoke'
         # [remote server] sun.reflect.DelegatingMethodAccessorImpl():-1:in `invoke'
         # [remote server] java.lang.reflect.Method():-1:in `invoke'
         # [remote server] org.openqa.selenium.support.events.EventFiringWebDriver$EventFiringWebElement$1(EventFiringWebDriver.java):331:in `invoke'
         # [remote server] com.sun.proxy.$Proxy2():-1:in `getText'
         # [remote server] org.openqa.selenium.support.events.EventFiringWebDriver$EventFiringWebElement(EventFiringWebDriver.java):381:in `getText'
         # [remote server] sun.reflect.GeneratedMethodAccessor4():-1:in `invoke'
         # [remote server] sun.reflect.DelegatingMethodAccessorImpl():-1:in `invoke'
         # [remote server] java.lang.reflect.Method():-1:in `invoke'
         # [remote server] org.openqa.selenium.remote.server.KnownElements$1(KnownElements.java):63:in `invoke'
         # [remote server] com.sun.proxy.$Proxy3():-1:in `getText'
         # [remote server] org.openqa.selenium.remote.server.handler.GetElementText(GetElementText.java):29:in `call'
         # [remote server] org.openqa.selenium.remote.server.handler.GetElementText(GetElementText.java):1:in `call'
         # [remote server] java.util.concurrent.FutureTask():-1:in `run'
         # [remote server] org.openqa.selenium.remote.server.DefaultSession$1(DefaultSession.java):169:in `run'
         # [remote server] java.util.concurrent.ThreadPoolExecutor():-1:in `runWorker'
         # [remote server] java.util.concurrent.ThreadPoolExecutor$Worker():-1:in `run'
         # [remote server] java.lang.Thread():-1:in `run'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/remote/response.rb:52:in `assert_ok'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/remote/response.rb:15:in `initialize'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/remote/http/common.rb:59:in `new'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/remote/http/common.rb:59:in `create_response'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/remote/http/default.rb:66:in `request'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/remote/http/common.rb:40:in `call'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/remote/bridge.rb:640:in `raw_execute'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/remote/bridge.rb:618:in `execute'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/remote/bridge.rb:529:in `getElementText'
         # /var/lib/gems/2.2.0/gems/selenium-webdriver-2.45.0/lib/selenium/webdriver/common/element.rb:108:in `text'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/selenium/node.rb:4:in `visible_text'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/element.rb:61:in `block in text'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/base.rb:80:in `synchronize'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/element.rb:57:in `text'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/query.rb:44:in `matches_filters?'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/result.rb:27:in `block in initialize'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/result.rb:27:in `select'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/result.rb:27:in `initialize'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/query.rb:118:in `new'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/query.rb:118:in `block in resolve_for'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/base.rb:80:in `synchronize'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/query.rb:106:in `resolve_for'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/finders.rb:32:in `block in find'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/base.rb:84:in `synchronize'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/node/finders.rb:30:in `find'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/session.rb:676:in `block (2 levels) in <class:Session>'
         # /var/lib/gems/2.2.0/gems/capybara-2.4.4/lib/capybara/dsl.rb:51:in `block (2 levels) in <module:DSL>'
         # ./spec/support/context_helper.rb:6:in `block in we_are_back_on_the_service_app'
         # ./spec/support/context_helper.rb:6:in `we_are_back_on_the_service_app'
         # ./spec/01_authenticate_spec.rb:13:in `block (2 levels) in <top (required)>'
     # ./spec/support/context_helper.rb:6:in `we_are_back_on_the_service_app'
     # ./spec/01_authenticate_spec.rb:13:in `block (2 levels) in <top (required)>'

Finished in 1 minute 11.68 seconds (files took 0.90381 seconds to load)
2 examples, 1 failure

Failed examples:

rspec ./spec/01_authenticate_spec.rb:4 # Authentication Can log in to the service app, using the auth app
```

**Comment by ansonK** [comment]
Other than that lgtm

**Comment by oskarpearson** [comment]
Capturing my perspective of a chat with Anson on Slack:

The failure of the browser above appears to be a valid failure from my side. It may not have a difference in appearance, but clicking the item appears to fail because of visibility, javascript, or differences in the way that Capistrano works on the different browsers.

We're going to leave this test as failing for now on that browser - because fixing that particular item is out of scope for the agile story I'm working on. It also shows how failures would be raised, which is useful given that the smoke tests are still very much in an initial state.

We'll have a succession of smoke-test stories coming up for things like like the below. The above error will thus become part of that series of stories. The stories will also sit more with the dev team than with myself:
- Increasing the list of browsers we test against to match the project requirements
- Improving test coverage of the tests.

**Pull request merged**